matchbox_socket/src/error.rs

```diff
@@ -1,9 +1,13 @@
-use crate::webrtc_socket::error::SignalingError;
+use crate::webrtc_socket::error::{MessageSendError, SignalingError};
 
-/// Errors that can happen when using Matchbox.
+/// Errors that can happen when using Matchbox sockets.
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
     /// An error occurring during the signaling loop.
     #[error("An error in the signaling loop: {0}")]
     Signaling(#[from] SignalingError),
+
+    /// An error occurring during the messaging loop.
+    #[error("An error in the message loop")]
+    Messaging(#[from] MessageSendError),
```

Comment:
I'm wondering if we are leaking implementation details here? Do the user need to know/care from what "loop" the error originated?

Unless I misremember, this is now returned by the future you get when creating a socket, which we've dubbed `MessageLoopFuture`, so in that sense, all our errors are `MessageLoop` errors from the user point of view.

Maybe we could flatten the struct or rename it so it matches the rest of the public API?

Reply:
That sounds good but I'm not sure how or what approach you'd like. Perhaps best as a future revision or suggestion to the PR?

Reply:
`run_socket` can return a `SignalingError` or a `MessageSendError` now, since we propagate the Message Loop error up now. I think the current implementation makes more sense, and it is valuable to know where the error occurred?

<img width="1292" alt="image" src="https://github.com/johanhelsing/matchbox/assets/48108917/511785dd-e41a-4650-bd59-ca08e2293be5">

Reply:
As I see it, the value in returning a specific error, is that the user can handle it appropriately. That is something like this:

```rust
match err {
    SignalingConnectionFailed => // ask user to check internet connection, and try again
    SignalingCommunication => // (unknown format, socket messages etc) apologize to the user for the infrastructure being broken
    SignalingServerDisconnected => // tell you were kicked by the server, or the server shut down etc.
}
```

If we expose what loop the error occurred in, we can't change the implementation without breaking the API, for instance before we just had one big loop, and there was no distinction between signaling and message loop.

If we want debug information, I think we should just use tracing

Reply:
I think I satisfied your request; the error for `run_socket` is now a flattened `SignalingError`, and then reformed to a `crate::Error` when the socket is built.

<img width="1008" alt="image" src="https://github.com/johanhelsing/matchbox/assets/48108917/d16f9d35-edca-454f-a1f4-741ee58ac009">

The crate Error seems representative of your ask:

<img width="1314" alt="image" src="https://github.com/johanhelsing/matchbox/assets/48108917/de8dadea-85b9-44af-8372-52459d611abf">